students/cindywaldron/lesson5/mailroom.py

```diff
@@ -0,0 +1,109 @@
+    donation_amount = 'Enter donation amount(enter Q to main menu):'
+    response = input(full_name)
+    # display all donors' names if input is 'list'
+    while response == 'list':
+        display_donors()
+        response = input(full_name)
+    # if input is not 'Q' to quit
+    if response.upper() != 'Q':
+        amount_input = input(donation_amount)
+        if amount_input.upper() != 'Q':
+            try:
+                amount = float(amount_input)
+                # go through each donor record
+                for name, donation in donors.items():
+                    # existing donor
+                    if response == name:
+                        donation[0] = donation[0] + amount
+                        donation[1] = donation[1] + 1
+                        donation[2] = donation[2]/donation[1]
+                        break
+                else:
+                    # new donor
+                    donors[response] =[amount, 1, amount]
+                # thank you email
+                print(write_letter(response, amount))
+            except ValueError:
+                print('Input must be a float.  try again')
+
+
+def send_letters():
+    """Send letters to all donors"""
+    for k, v in donors.items():
+        letter = write_letter(k, v[0])
+       # strip ','
+        filename = k.replace(',','')
+       # replace space with '_'
+        filename = filename.replace(' ', '_')
+        # open file to write
+        with open('{}-{}{}'.format(filename, str(date.today()),'.txt'), 'w') as thefile:
+            thefile.write(letter)
+
+def create_report():
+    """Generate report"""
+    print("--------------------------------------------------------------")
+    print("{:20} | {:10} | {:5} | {:10}".format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift'))
+    print("--------------------------------------------------------------")
+    for k, v in sorted(donors.items(), key=get_key, reverse=True):
+        a_row = '{:20}  $ {:>10.2f}  {:>10d}  $ {:>11.2f}'.format(k, v[0],v[1],v[2])
+        print(a_row)
+
+def display_donors():
+    """
+    Display donor names
+    """
+    new_list = [ donor for donor in donors.keys()]
+    print(*new_list, sep = "\n")
+
+
+def quit():
+    return 'exit menu'
+
+def display_main_menu():
+    """Display main menu"""
+    print('Choose an action:')
+    options = ['1. Send Thank You', '2. Create a Report', '3. Send letters to everyone', '4. Quit']
+    option_str = '\n'.join(['\t'+item for item in options])
+    print(option_str)
+
+
+if __name__ == '__main__':
+    choice = ''
+    selection = 'Select an option (1, 2, 3, or 4) ===>'
+    switch_function_dict = {'1': send_thank_you, '2': create_report, '3': send_letters, '4': quit}
+    while True:
+        # Display main menu
+        display_main_menu()
+        try:
+            choice = input(selection)
+            if switch_function_dict[choice]() == 'exit menu':
+                break
+        except KeyError:
```

Comment:
it i best to wrap try/except around code you expect to throw that exception (in this case lines 106, 107)